File: src/test/java/com/bbcnews/automation/scripts/BBCNewsStatsTest.java
package com.bbcnews.automation.scripts;

import com.aventstack.extentreports.ExtentTest;
import com.bbcnews.automation.base.CharlesProxy;
import com.bbcnews.automation.commonfunctions.CommonFunctions;
import com.bbcnews.automation.pageobjects.*;
import com.bbcnews.automation.testutils.AppiumStart;
import com.bbcnews.automation.testutils.ExtenttestReport;
import io.appium.java_client.MobileElement;
import io.appium.java_client.android.AndroidDriver;
import io.appium.java_client.pagefactory.AppiumFieldDecorator;
import io.appium.java_client.remote.MobileCapabilityType;
import io.qameta.allure.Severity;
import io.qameta.allure.SeverityLevel;
import io.qameta.allure.Story;
import org.openqa.selenium.ScreenOrientation;
import org.openqa.selenium.remote.DesiredCapabilities;
import org.openqa.selenium.support.PageFactory;
import org.testng.Assert;
import org.testng.ITestResult;
import org.testng.annotations.AfterMethod;
import org.testng.annotations.AfterTest;
import org.testng.annotations.BeforeTest;
import org.testng.annotations.Test;

import java.io.File;
import java.io.IOException;
import java.net.URL;
import java.text.SimpleDateFormat;
import java.util.Date;

public class BBCNewsStatsTest {

        private AndroidDriver<MobileElement> androidDriver;
        private DesiredCapabilities capabilities = new DesiredCapabilities();
        private String Deviceos_Name ;
        private String Device_id;
        private String Device_Name;
        private String App_Path;
        private String Appium_Port;
        public File file;
        ExtentTest test;

        private HomePageObject homePageObject;
        private MyNewsPageObject myNewsPageObject;
        private BasePageObject basePageObject;
        private VidoePageObject vidoePageObject;
        private PopularPageObject popularPageObject;

        ExtenttestReport extenttestReport = new ExtenttestReport();
        AppiumStart appiumStart = new AppiumStart();
        CharlesProxy charlesProxy = new CharlesProxy();

        private static Date curDate = new Date();
        private static SimpleDateFormat format = new SimpleDateFormat("yyyy-MM-dd");
        private static String DateToStr = format.format(curDate);
        public static String workingDirectory = System.getProperty("user.dir");
        private static String screenshotpath = workingDirectory + "/Screenshots/";

        @BeforeTest
        public void RunTest() throws Exception {

        try {

            readDeviceDetailsCommandPrompt();
            charlesProxy.startCharles();
            charlesProxy.charlesSessionStart();
            setUP();
            LaunchBBCNews();
        } catch (Exception e) {
            e.printStackTrace();
        }

    }

        public void readDeviceDetailsCommandPrompt() {

        try {
            Deviceos_Name = System.getProperty("DeviceOS");
            Device_id = System.getProperty("DeviceID");
            Device_Name = System.getProperty("DeviceName");
            App_Path = System.getProperty("AppPath");
            Appium_Port = System.getProperty("AppiumPort");
            System.out.println("Passed The Device OS is " + Deviceos_Name);
            System.out.println("Passed The Device ID is " + Device_id);
            System.out.println("Passed The Device Name is " + Device_Name);
            System.out.println("Passed The Appium port is " + Appium_Port);
            System.out.println("Passed The Application path  is " + App_Path);
        } catch (Exception e) {
            e.printStackTrace();

        }

    }

        public void setUP() {
        try
        {
            //  appiumStart.startAppium(Integer.parseInt(Appium_Port));
            String appium_url = "http://127.0.0.1:"+Appium_Port+"/wd/hub";
            System.out.println("Appium Server Address : - " + appium_url);
            capabilities = new DesiredCapabilities();
            capabilities.setCapability(MobileCapabilityType.UDID, Device_id);
            capabilities.setCapability(MobileCapabilityType.DEVICE_NAME, "bbcnews");
            capabilities.setCapability(MobileCapabilityType.AUTOMATION_NAME, "UiAutomator2");
            capabilities.setCapability(MobileCapabilityType.PLATFORM_NAME, "Android");
            capabilities.setCapability("appiumversion", "1.8.1");
            capabilities.setCapability("app",App_Path); //"/Users/ramakh02/Desktop/tools/APK/BBCNews-5.5.0.35.apk");
            //it's not mandatory to pass OS version of the device
            // capabilities.setCapability(MobileCapabilityType.PLATFORM_VERSION,Deviceos_Name);
            capabilities.setCapability("appPackage", "bbc.mobile.news.uk.internal" );
            capabilities.setCapability("appActivity","bbc.mobile.news.v3.app.TopLevelActivity");
            // capabilities.setCapability(MobileCapabilityType.UNEXPECTED_ALERT_BEHAVIOUR, true);
            // capabilities.setCapability("autoAcceptAlerts", true);
            capabilities.setCapability("--session-override",true);
            androidDriver = new AndroidDriver<MobileElement>(new URL(appium_url),capabilities);
        } catch (Exception e) {
        }
    }

        public void LaunchBBCNews() {
        try {
            homePageObject = new HomePageObject();
            PageFactory.initElements(new AppiumFieldDecorator(androidDriver), homePageObject);

            myNewsPageObject = new MyNewsPageObject();
            PageFactory.initElements(new AppiumFieldDecorator(androidDriver), myNewsPageObject);

            basePageObject = new BasePageObject();
            PageFactory.initElements(new AppiumFieldDecorator(androidDriver), basePageObject);

            vidoePageObject = new VidoePageObject();
            PageFactory.initElements(new AppiumFieldDecorator(androidDriver), vidoePageObject);

            popularPageObject = new PopularPageObject();
            PageFactory.initElements(new AppiumFieldDecorator(androidDriver), popularPageObject);

            CommonFunctions.emptyFolder(screenshotpath);

            extenttestReport.createrReportHive("SmokeTest", Deviceos_Name, Device_Name, Device_id);

            androidDriver.context("NATIVE_APP");
            file = new File(screenshotpath);
            String screenshot = file.getAbsolutePath();
            System.out.println("The ScreenShot Path is " + screenshot);

            ScreenOrientation orientation = androidDriver.getOrientation();
            if(orientation.equals(ScreenOrientation.LANDSCAPE))
            {
                androidDriver.rotate(ScreenOrientation.PORTRAIT);
            }


        } catch (NullPointerException e) {
            e.printStackTrace();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @Test(priority = 1, description = "launching the app and start the Charles ")
    @Story("Home")
    @Severity(SeverityLevel.CRITICAL)
    public void testOpenNewsApp() throws Exception {
        try {


            extenttestReport.tapButton(androidDriver,basePageObject.okbutton,false);
            extenttestReport.tapButton(androidDriver,basePageObject.nothanksbutton,false);


        } catch (Exception e) {
            e.printStackTrace();
        }
    }


    @Test(priority = 2, description = "Test to navigate around app for generate stats")
    @Story("Home")
    @Severity(SeverityLevel.CRITICAL)
    public void testCheckHomePage(){
        try {
            extenttestReport.startTest("HomePage", "Checking the HomePage", "Smoke");
            extenttestReport.tapButton(androidDriver,basePageObject.mynews,false);
            extenttestReport.tapButton(androidDriver,basePageObject.popular,false);
            extenttestReport.tapButton(androidDriver,basePageObject.video,false);

        } catch (Exception e) {
            e.printStackTrace();
        }
    }


    @Test(priority = 3, description = "Test to stop the charles recording ")
    public void testStopcharlesRecord() throws Exception {

        try {
            charlesProxy.stopCharlesSession();
            Thread.sleep(8000);
        }catch (AssertionError e)
        {
            throw e;
        }
    }


    @Test(priority = 4, description = "Test to download the charles recording session")
    public void testDownloadcharlesRecord() throws Exception {

        try {
            charlesProxy.downloadCharlesSession();
            Thread.sleep(8000);
        }catch (AssertionError e)
        {
            throw e;
        }
    }

    @Test(priority = 5, description = "Test to convert the charles session to CSV format")
    public void testConvertCharlesSessiontoCSV() throws Exception {

        try {
            charlesProxy.converttoCSV();
            Thread.sleep(8000);
            charlesProxy.stopCharles();
        }catch (AssertionError e)
        {
            throw e;
        }
    }

    @AfterMethod
    public void getResult(ITestResult result)
    {
        try {
            extenttestReport.getTestResult(androidDriver,result);
        }catch (IOException e)
        {}

    }
    @AfterTest
    public void tearDown()
    {
        extenttestReport.publishReport();
        androidDriver.closeApp();
        androidDriver.quit();

    }


}

File: src/test/java/com/bbcnews/automation/scripts/BBCNewsStatsTest.kt
package com.bbcnews.automation.scripts

import com.aventstack.extentreports.ExtentTest
import com.bbcnews.automation.testutils.CharlesProxy
import com.bbcnews.automation.commonfunctions.CommonFunctionKotlin
import com.bbcnews.automation.pageobjects.*
import com.bbcnews.automation.testutils.Testutility
import io.appium.java_client.MobileElement
import io.appium.java_client.android.AndroidDriver
import io.appium.java_client.pagefactory.AppiumFieldDecorator
import io.appium.java_client.remote.MobileCapabilityType
import io.qameta.allure.Severity
import io.qameta.allure.SeverityLevel
import io.qameta.allure.Story
import org.openqa.selenium.ScreenOrientation
import org.openqa.selenium.remote.DesiredCapabilities
import org.openqa.selenium.support.PageFactory
import org.testng.ITestResult
import org.testng.annotations.AfterMethod
import org.testng.annotations.AfterTest
import org.testng.annotations.BeforeTest
import org.testng.annotations.Test

import java.io.File
import java.io.IOException
import java.net.URL
import java.text.SimpleDateFormat
import java.util.Date

class BBCNewsStatsTest {

    private var capabilities = DesiredCapabilities()
    private var deviceosName: String? = null
    private var deviceid: String? = null
    private var deviceName: String? = null
    private var appPath: String? = null
    private var appiumPort: String? = null
    private lateinit var file: File

    private var commonFunctionKotlin = CommonFunctionKotlin()
    private var testutility = Testutility()

    private lateinit var homePageObject: HomePageObject
    private lateinit var androidDriver: AndroidDriver<MobileElement>
    private lateinit var myNewsPageObject: MyNewsPageObject
    //private lateinit var basePageObject: BasePageObject
    private lateinit var vidoePageObject: VideoPageObjects
    private lateinit var popularPageObject: PopularPageObjects
    private lateinit var basePageObjectModel: BasePageObject
    private lateinit var commonpageobjects: CommonPageObjects

    internal var charlesProxy = CharlesProxy()
    var statsTestData = StatsTestData()


    @BeforeTest
    @Throws(Exception::class)
    fun RunTest() {

        try {

            readDeviceDetailsCommandPrompt()
            charlesProxy.startCharles()
            setUP()
            initialiseobjects()
        } catch (e: Exception) {
            e.printStackTrace()
        }

    }

    fun readDeviceDetailsCommandPrompt() {

        try {
            deviceosName = System.getProperty("DeviceOS")
            deviceid = System.getProperty("DeviceID")
            deviceName = System.getProperty("DeviceName")
            appPath = System.getProperty("AppPath")
            appiumPort = System.getProperty("AppiumPort")
            println("Passed The Device OS is $deviceosName")
            println("Passed The Device ID is $deviceid")
            println("Passed The Device Name is $deviceName")
            println("Passed The Appium port is $appiumPort")
            println("Passed The Application path  is $appPath")
        } catch (e: Exception) {
            e.printStackTrace()

        }

    }

    private  fun setUP() {
        try {
            //  appiumStart.startAppium(Integer.parseInt(Appium_Port));
            val appiumurl = "http://127.0.0.1:$appiumPort/wd/hub"
            println("Appium Server Address : - $appiumurl")
            capabilities = DesiredCapabilities()
            capabilities.setCapability(MobileCapabilityType.UDID, deviceid)
            capabilities.setCapability(MobileCapabilityType.DEVICE_NAME, "bbcnews")
            capabilities.setCapability(MobileCapabilityType.AUTOMATION_NAME, "UiAutomator2")
            capabilities.setCapability(MobileCapabilityType.PLATFORM_NAME, "Android")
            capabilities.setCapability("appiumversion", "1.8.1")
            capabilities.setCapability("app", appPath) //"/Users/ramakh02/Desktop/tools/APK/BBCNews-5.5.0.35.apk");
            //it's not mandatory to pass OS version of the device
            // capabilities.setCapability(MobileCapabilityType.PLATFORM_VERSION,Deviceos_Name);
            capabilities.setCapability("appPackage", "bbc.mobile.news.uk.internal")
            capabilities.setCapability("appActivity", "bbc.mobile.news.v3.app.TopLevelActivity")
            // capabilities.setCapability(MobileCapabilityType.UNEXPECTED_ALERT_BEHAVIOUR, true);
            // capabilities.setCapability("autoAcceptAlerts", true);
            capabilities.setCapability("--session-override", true)
            capabilities.setCapability("ignoreUnimportantViews",true)
            androidDriver = AndroidDriver(URL(appiumurl), capabilities)
        } catch (e: Exception) {
        }
    }

    private fun initialiseobjects() = try {
        homePageObject = HomePageObject()
        PageFactory.initElements(AppiumFieldDecorator(androidDriver), homePageObject)

        myNewsPageObject = MyNewsPageObject()
        PageFactory.initElements(AppiumFieldDecorator(androidDriver), myNewsPageObject)

        basePageObjectModel = BasePageObject()
        PageFactory.initElements(AppiumFieldDecorator(androidDriver), basePageObjectModel)

        vidoePageObject = VideoPageObjects()
        PageFactory.initElements(AppiumFieldDecorator(androidDriver), vidoePageObject)

        popularPageObject = PopularPageObjects()
        PageFactory.initElements(AppiumFieldDecorator(androidDriver), popularPageObject)

        testutility.emptyFolder(screenshotpath)


        commonFunctionKotlin.createrReportHive("Regression", deviceName.toString(), deviceid.toString())


        androidDriver.context("NATIVE_APP")
        file = File(screenshotpath)
        val screenshot = file.absolutePath
        println("The ScreenShot Path is $screenshot")

        val orientation = androidDriver.orientation
        if (orientation != ScreenOrientation.LANDSCAPE) {
            androidDriver.rotate(ScreenOrientation.PORTRAIT)
        }
        else { }
    } catch (e: NullPointerException) {
        e.printStackTrace()
    } catch (e: Exception) {
        e.printStackTrace()
    }

    @Test(priority = 1, description = "launching the app and start the Charles ")
    @Story("Home")
    @Severity(SeverityLevel.CRITICAL)
    @Throws(Exception::class)
    fun testOpenNewsApp() {
        try {


            commonFunctionKotlin.tapButton(androidDriver, basePageObjectModel.okbutton, false)
            commonFunctionKotlin.tapButton(androidDriver, basePageObjectModel.nothanksbutton, false)
            commonFunctionKotlin.tapButton(androidDriver, basePageObjectModel.menubutton, false)
            commonFunctionKotlin.tapButton(androidDriver, basePageObjectModel.internalsettings, false)
            commonFunctionKotlin.tapButton(androidDriver, basePageObjectModel.cpscontent, false)
            commonFunctionKotlin.tapButton(androidDriver, basePageObjectModel.trevortest, false)
            commonFunctionKotlin.tapButton(androidDriver, basePageObjectModel.navigate_back, false)
            commonFunctionKotlin.tapButton(androidDriver, basePageObjectModel.reloadButton, false)
            charlesProxy.startcharlesSession()

        } catch (e: Exception) {
            e.printStackTrace()
        }

    }


    @Test(priority = 2, description = "Test to navigate around app for generate stats")
    @Story("Home")
    @Severity(SeverityLevel.CRITICAL)
    fun testCheckHomePage() {
        try {
            //  androidDriver.runAppInBackground(Duration.ofSeconds(15));

            //  androidDriver.currentActivity();
            commonFunctionKotlin.startTest("HomePage", "Checking the HomePage", "Smoke")
            commonFunctionKotlin.tapButton(androidDriver, basePageObjectModel.topstories, false)
            commonFunctionKotlin.tapButton(androidDriver,basePageObjectModel.mynews, false)
            commonFunctionKotlin.tapButton(androidDriver, basePageObjectModel.popular, false)
            commonFunctionKotlin.tapButton(androidDriver, basePageObjectModel.video, false)
            commonFunctionKotlin.tapButton(androidDriver, basePageObjectModel.searchbutton, false)

        } catch (e: Exception) {
            e.printStackTrace()
        }

    }


    @Test(priority = 3, description = "Test to stop the charles recording ")
    @Throws(Exception::class)
    fun testStopcharlesRecord() {

        try {
            charlesProxy.stopCharlesSession()
            Thread.sleep(2000)
        } catch (e: AssertionError) {
            throw e
        }

    }


    @Test(priority = 4, description = "Test to download the charles recording session")
    @Throws(Exception::class)
    fun testDownloadcharlesRecord() {

        try {
            charlesProxy.downloadCharlesSession()
            Thread.sleep(2000)
        } catch (e: AssertionError) {
            throw e
        }

    }

    @Test(priority = 5, description = "Test to convert the charles session to CSV format")
    @Throws(Exception::class)
    fun testConvertCharlesSessiontoCSV() {

        try {
            charlesProxy.converttoCSV()
            Thread.sleep(2000)
            charlesProxy.stopCharles()
        } catch (e: AssertionError) {
            throw e
        }

    }


    @Test(priority = 6, description = "Test to check the compared the downloaded stats")
    fun testBBVNewsBasicStats() {
        commonFunctionKotlin.startTest("BasicStats", "Test to check the compared the downloaded stats", "Stat's")
        try {
            statsTestData.comapre_StatsData(statsTestData.csvFile, statsTestData.BBCnewsBasicstats)
        } catch (e: InterruptedException) {
            e.printStackTrace()
        } catch (e: IOException) {
            e.printStackTrace()
        }

    }


    @Test(priority = 7, description = "Test to check the TopStores Page downloaded stats")
    fun testCheckTopStoresStats() {
        commonFunctionKotlin.startTest("TopStores", "Test to check the TopStores Page downloaded stats", "TopStoriesStat's")
        try {
            statsTestData.comapre_StatsData(statsTestData.csvFile, statsTestData.topstores)
        } catch (e: InterruptedException) {
            e.printStackTrace()
        } catch (e: IOException) {
            e.printStackTrace()
        }

    }

    @Test(priority = 8, description = "Test to check the MyNews Page downloaded stats")
    fun testCheckMyNewsStats() {
        commonFunctionKotlin.startTest("MyNews", "Test to check the MyNews Page downloaded stats", "MyNews Stat's")
        try {
            statsTestData.comapre_StatsData(statsTestData.csvFile, statsTestData.mynews)
        } catch (e: InterruptedException) {
            e.printStackTrace()
        } catch (e: IOException) {
            e.printStackTrace()
        }

    }

    @Test(priority = 9, description = "Test to check the Popular Page downloaded stats")
    fun testCheckPopularPageStats() {
        commonFunctionKotlin.startTest("Popular", "Test to check the Popular Page downloaded stats", "Popular Stat's")
        try {
            statsTestData.comapre_StatsData(statsTestData.csvFile, statsTestData.popularpage)
        } catch (e: InterruptedException) {
            e.printStackTrace()
        } catch (e: IOException) {
            e.printStackTrace()
        }

    }

    @Test(priority = 10, description = "Test to check the Video Page downloaded stats")
    fun testCheckVideoPageStats() {
        commonFunctionKotlin.startTest("Video", "Test to check the Video Page downloaded stats", "Video Stat's")
        try {
            statsTestData.comapre_StatsData(statsTestData.csvFile, statsTestData.videopage)
        } catch (e: InterruptedException) {
            e.printStackTrace()
        } catch (e: IOException) {
            e.printStackTrace()
        }

    }

    @Test(priority = 11, description = "Test to check the Search Page downloaded stats")
    fun testCheckPopularStats() {
        commonFunctionKotlin.startTest("Search", "Test to check the Search Page downloaded stats", "Search Stat's")
        try {
            statsTestData.comapre_StatsData(statsTestData.csvFile, statsTestData.searchstats)
        } catch (e: InterruptedException) {
            e.printStackTrace()
        } catch (e: IOException) {
            e.printStackTrace()
        }

    }


    @AfterMethod
    fun getResult(result: ITestResult) {
        try {
            commonFunctionKotlin.getTestResult(androidDriver, result)
        } catch (e: IOException) {
        }

    }

    @AfterTest
    fun tearDown() {
        commonFunctionKotlin.publishReport()
        androidDriver.closeApp()
        androidDriver.quit()

    }

    companion object {

        private val curDate = Date()
        private val format = SimpleDateFormat("yyyy-MM-dd")
        private val DateToStr = format.format(curDate)
        var workingDirectory = System.getProperty("user.dir")
        private val screenshotpath = "$workingDirectory/Screenshots/"
    }


}


